keel-core/src/main/kotlin/com/netflix/spinnaker/keel/services/ApplicationService.kt

```diff
@@ -277,18 +276,18 @@ class ApplicationService(
     }
 
     val artifactSummaries = deliveryConfig.artifacts.map { artifact ->
+      // We calculate a bunch of data for each artifact, so pre-load the bulk of it here and pass that info into
+      //  later functions to save redundant database calls.
       val artifactVersions = repository.artifactVersions(artifact, limit)
+      val artifactVersionInfoInEnvironment: Map<String, List<StatusInfoForArtifactInEnvironment>> = deliveryConfig.environments.map { env ->
+        env.name to repository.getVersionInfoInEnvironment(deliveryConfig, env.name, artifact)
+      }.toMap()
```

Comment:
```suggestion
      val artifactVersionInfoInEnvironment = deliveryConfig.environments.associate { env ->
        env.name to repository.getVersionInfoInEnvironment(deliveryConfig, env.name, artifact)
      }
```

Reply:
P.S. Seems so obvious to load all the data upfront in hindsight, doesn't it? 🙂  Thanks for making this optimization! ❤️

Reply:
It took me quite a lot of staring to realize how much this would save in db calls! I'm so thrilled to open this PR.